### PR TITLE
Many changes 

### PR DESCRIPTION
**Pull request by dougg0k**
Hello!

A few days ago I looked at the project and got interested in it, seemed to be a better option than bee-queue. And I looked through the issues and made this PR.

But this PR contains more things than that. 

- It seem to address this, at least the lua are sent to dist. https://github.com/conveyor-mq/conveyor-mq/issues/114
- I completely addressed this. https://github.com/conveyor-mq/conveyor-mq/issues/68
- I migrated the project from `npm` to `pnpm` which are very fast. and is the one I have been using in my projects. After trying to install packages using recent `npm` version, it that was complaining / giving errors.
- I upgraded most packages, with some exceptions that has breaking changes, but they were partially upgraded, to the point that compatibility was still there.
- I fixed type errors with and without additional checks.
- Imports were re-ordered alphabetically.

It could be worth upgrading to `ioredis` latest version, seems that is more stable but there are breaking changes. As for the `p-` starting dependencies, it is in the latest version that is usable in non-ESM projects. While the rest are up-to-date.

Edit: I upgraded ioredis to v5, only small breaking changes, mostly in the typing.

Edit2: There is a test that randomly fails. Been happening before I started the changes though. Sometimes I get consecutives test run all successful and sometimes consecutives 1 fail with that one below.
```
createOrchestrator › createOrchestrator marks orphaned task as processing
> 78 |     expect(await isTaskStalled({ taskId, queue, client })).toBe(true);
```

It's up to you to accept or not. All tests seem to be passing. 

@jasrusable

### CODE REVIEW

**Comment by jasrusable** [comment]
Thanks for this PR! I'll take a look through it when I have some time.

**Comment by dougg0k** [comment]
Hey @jasrusable are you able to get this merged in?

**Comment by dougg0k** [comment]
@jasrusable ? I wanted to use the project with these changes.

**Comment by jasrusable** [comment]
>Edit2: There is a test that randomly fails. Been happening before I started the changes though. Sometimes I get consecutives test run all successful and sometimes consecutives 1 fail with that one below.

I'll take a look at this test and see if I can spot why it might be flaky.

**Comment by dougg0k** [comment]
> Hey @dougg0k :) I took a quick look through the PR, for the most part I think it looks great! Thanks again.
> 
> Two main things though:
> 
> 1. I'd prefer to stick with vanilla `npm` for the sake of its ubiquity for now.
> 2. I'd prefer to use `date-fns` over `datejs` due to its simplicity.
> 
> Would you mind please implementing those changes?

- `npm` gives problems on install I would need to use legacy peer flag to make it work.
- `datejs` are much smaller and has the same features as moment, while being immutable. https://github.com/conveyor-mq/conveyor-mq/issues/68#issuecomment-1236414709

**Comment by jasrusable** [comment]
What errors does npm give on installation? Can you share the error logs?

**Comment by dougg0k** [comment]
Nevermind, it did while I was working on it.

**Comment by dougg0k** [comment]
https://github.com/raineorshine/npm-check-updates Much better and easier to update  libs while not having the bloat of dependabot and attempts of lib updates that goes to ESM only. You can just `--reject` those.

**Comment by jasrusable** [comment]
Great, then last few things:
1. Switch `datejs` to `date-fns`
2. Fix failing tests

**Comment by dougg0k** [comment]
May I ask why keep it compatible with older versions than 14 LTS?

File: package.json [hunk 1]
```diff
@@ -36,39 +36,37 @@
...
-    "set-interval-async": "^1.0.32",
-    "uuid": "^8.1.0"
+    "dayjs": "1.11.5",
+    "debug": "4.3.4",
+    "ioredis": "5.2.3",
+    "p-queue": "6.6.2",
+    "p-timeout": "4.1.0",
+    "rate-limiter-flexible": "2.3.12",
+    "set-interval-async": "3.0.2",
+    "uuid": "9.0.0"
```
Comment (falkenhawk) [inline package.json:46]: I don't think `dependencies` in "library"-like packages should be fixed to exact versions. Could we still have carets here, please?

Comment (dougg0k) [reply package.json:46]: Isn't that the whole point of have proper versions and to keep compatibility? Caret may install a different version than what you see.

Comment (dougg0k) [reply package.json:46]: caret are minor + patch. Maybe no breaking changes if every lib are following the semantics.

**Comment by dougg0k** [comment]
Not sure if the jest issue are fixed, because it's in older node versions. Might not be compatible. It would be good to run automatically to test.

**Comment by dougg0k** [comment]
All of the tests are passing / green here, might be the compatibility with older versions. Not sure I want to work on that part. If you or anyone want to work on it.

**Comment by dougg0k** [comment]
v18 LTS will release soon, Oct 25, do you agree that make sense to keep support only to v14+?

**Comment by dougg0k** [comment]
Can you re-run the tests? @jasrusable

**Comment by dougg0k** [comment]
Now it's one incompability with `set-interval-async` in nodejs v12 and that test I said before that was failing randomly, so much they have all succeeded in some runs but not in others. 

Can I drop v12 from tests, too?

Will you take a look at that failing test? @jasrusable 
It seems there were 2, but here I only ever seen 1 failing randomly.

**Comment by jasrusable** [comment]
> Now it's one incompability with `set-interval-async` in nodejs v12 and that test I said before that was failing randomly, so much they have all succeeded in most runs but not in one.
> 
> Can I drop v12 from tests, too?
> 
> Will you take a look at that failing test? @jasrusable It seems there were 2, but here I only ever seen 1 failing randomly.

I’ll take a look at the remaining failing tests.

**Comment by dougg0k** [comment]
Cool, thanks.

**Comment by jasrusable** [comment]
@dougg0k seems it's just a single test that fails intermittently. Can you try increasing the sleep time to see if that fixes the test? 
image

**Comment by dougg0k** [comment]
Yeah, that is the test I mentioned, I already tried before increasing, didnt work.

**Comment by dougg0k** [comment]
I mean, if passes and fails randomly with the same value, must be a bug somewhere. 
Maybe if there were a way after the creation of the orchestrator to know if the whatever it is doing was done or not. Rather than be independant.

I saw some random fails of. After trying multiple runs.
```
 FAIL  src/tests/actions/acknowledge-task.test.ts
  ● acknowledgeTask › acknowledgeTask acknowledges task

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      37 |     expect(await isTaskStalled({ taskId: task.id, queue, client })).toBe(false);
      38 |     await sleep(50);
    > 39 |     expect(await isTaskStalled({ taskId: task.id, queue, client })).toBe(true);
```
It seems to be related to the same bug. Where it expect the task to be stalled but are not. 

@jasrusable

**Comment by dougg0k** [comment]
I tried running where one time succeeded and one failed. Any idea what could cause it to behave like this? @jasrusable 
`areTasksStalled`
<table>
  <tr>
    <td>Succeeded</td>
    <td>Failed</td>
  </tr>
  <tr>
    <td> console.log
    results - []

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - []

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [1]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["1bb5345d-46b9-4f0a-831a-38cd46bb871e"]

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [1]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["1bb5345d-46b9-4f0a-831a-38cd46bb871e"]

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [0]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["1bb5345d-46b9-4f0a-831a-38cd46bb871e"]

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [0]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["1bb5345d-46b9-4f0a-831a-38cd46bb871e"]

      at src/actions/are-tasks-stalled.ts:31:11</td>
</td>
<td>
  console.log
    results - []

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - []

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [1]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["56733eb8-1857-42e7-9b32-6a0dcd13b1d1"]

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - [0]

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - ["56733eb8-1857-42e7-9b32-6a0dcd13b1d1"]

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - []

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - []

      at src/actions/are-tasks-stalled.ts:31:11

  console.log
    results - []

      at src/actions/are-tasks-stalled.ts:30:11

  console.log
    stallingTasksIds - []

      at src/actions/are-tasks-stalled.ts:31:11


</td>
  </tr>
</table>

**Comment by jasrusable** [comment]
Strange, I cannot reproduce this failing test locally at all now, with either combination of Node 14, 16 or 18, Redis 4 or 5.

**Pull request merged**